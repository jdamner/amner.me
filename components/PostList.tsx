import React from "react";

/* Components */
import Title from "./Atoms/Title";
import ButtonLink from "./Atoms/Links/ButtonLink";

export default function PostList(
  props: React.DetailedHTMLProps<
    React.HTMLAttributes<HTMLDivElement>,
    HTMLDivElement
  > & { posts: { slug: string; title: string }[] },
) {
  const { posts, ...rest } = props;
  return (
    <div {...rest}>
      <Title>Blog Posts</Title>
      <ul>
        {posts &&
          posts.map((post, index) => (
            <li key={index} className="mb-1">
              <ButtonLink href={post.slug} key={index}>
                {post.title}
              </ButtonLink>
            </li>
          ))}
      </ul>
    </div>
  );
}
