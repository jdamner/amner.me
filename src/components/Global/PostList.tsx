import React from "react";
/* API */
import { motion, useReducedMotion } from "framer-motion"

/* Components */
import PostLink from "../Links/PostLink"

/* Types */
import type { Post } from "../../types/Post.type"
import type { MotionProps } from "framer-motion"

/**
 * Post List
 * 
 * @param {Post[]} posts
 * @returns JSX.Element
 */
export default function PostList({ posts }: { posts: Post[] }): React.JSX.Element {
  const reduceMotion: boolean = useReducedMotion()
  const PostLinks = posts && posts.map((post, index) => {
    const delay = index * 0.01
    const motionProps: MotionProps = {
      initial: { opacity: reduceMotion ? 1 : 0 },
      whileInView: { opacity: 1 },
      viewport: { once: true },
      transition: { delay, duration: 0.5 },
    }
    return <motion.div {...motionProps} key={index} className='break-inside-avoid-column'>
      <PostLink post={post} />
    </motion.div>
  })

  return (
    <div className='columns-1 xs:columns-2 md:columns-3 lg:columns-4 gap-2 space-y-2' id='post-list' >
      {PostLinks}
    </div >
  );
}