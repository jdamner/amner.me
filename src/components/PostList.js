import React from "react"

import PostListItem from "./PostListItem"

export default function PostList({ data }) {
    if (!data.allMarkdownRemark) {
        return null;
    }

    const { allMarkdownRemark } = data;
    const Posts = allMarkdownRemark.edges
    .filter(edge => !!edge.node.frontmatter.date) // You can filter your posts based on some criteria
    .map(edge => <PostListItem key={edge.node.id} post={edge.node} />)

    return Posts
}
