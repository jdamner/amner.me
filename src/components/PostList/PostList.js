import React from "react"
import { graphql, StaticQuery, Link } from "gatsby";
import { GatsbyImage, getImage } from "gatsby-plugin-image";
import { motion } from "framer-motion";

export default function PostList({ data }) {
  return (<StaticQuery
    query={graphql`
        query {
            allMarkdownRemark(sort: {fields: frontmatter___date, order: DESC}) {
              edges {
                node {
                  id
                  excerpt(pruneLength: 250)
                  frontmatter {
                    date
                    slug
                    thumbnail {
                      childImageSharp {
                        gatsbyImageData(
                          width: 200
                          placeholder: BLURRED
                          formats: [AUTO, WEBP, AVIF]
                        )
                      }
                    }
                    title
                  }
                }
              }
            }
          }
          `}
    render={data => {
      const { allMarkdownRemark } = data;
      var i = 0;
      const Posts = allMarkdownRemark.edges
        .filter(edge => !!edge.node.frontmatter.date) // You can filter your posts based on some criteria
        .map(edge => {
          i++;
          const delay = i * 0.1;
          const image = getImage(edge.node.frontmatter.thumbnail);

          return (<motion.div
            key={edge.node.id}
            initial={{ opacity: 0 }}
            whileInView={{ opacity: 1 }}
            transition={{ delay, duration: 0.5 }}
            className='grid-item'>
            <Link to={'/' + edge.node.frontmatter.slug}>
              {image ? (
                <GatsbyImage image={image} alt={edge.node.frontmatter.title} loading='lazy' />
              ) : null}
              <h3>
                {edge.node.frontmatter.title}
              </h3>
            </Link>
          </motion.div>)
        })
      return Posts
    }}
  />)


}