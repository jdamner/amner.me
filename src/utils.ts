/**
 * Helper functions
 */
import fs from 'fs'
import matter from 'gray-matter'
import { join } from 'path'
import { slug } from 'github-slugger'

import type { ReactNode } from 'react'
import type { MdFile } from './types'

/**
 * Makes an object JSON parseable
 * 
 * @param {any} obj The object to be parsed
 * @returns {any}
 */
export function makeJsonParseable<Type>(obj: Type): Type {
	return JSON.parse(JSON.stringify(obj)) as Type
}

/**
 * Reads a markdown file
 */
export async function readMdFile(filename: string, directory: string = 'content/'): Promise<MdFile> {
	const path = `${directory}/${filename}`
	const fullPath = `${process.cwd()}/${path}`
	const fileContents = fs.readFileSync(fullPath, 'utf8')
	return {
		slug: slug(filename.replace(/\.md$/, '')),
		...matter(fileContents),
	}
}

/**
 * Reads all slugs from a directory
 * 
 * @param {string} directory The directory to read from
 * 
 * @returns {string[]}
 */
export function getFilesInDir(directory: string): string[] {
	const fullPath = join(process.cwd(), directory)
	return fs.readdirSync(fullPath)
}

/**
 * Converts all headings in a markdown file to have IDs 
 * Converts 
 * ## Heading 1 
 * to become
 * ## Heading 1{#heading-1}
 * 
 * @param {string} markdown The markdown to convert
 * 
 * @returns {string}
 */
export function convertHeadingsToAnchorLinks(markdown: string): string {
	return markdown.replace(/(#{1,6})\s(.+)\n/gm, (match, p1, p2) => {
		return `${p1} ${p2}{#${slug(p2)}}\n`
	})
}


/**
 * Converts a react node to a slug
 * 
 * Usually the value is a string, but sometimes it contains
 * a `<code>` element, which is a react node. This function
 * converts the value to a string.
 */
export function makeSlug(value: ReactNode | ReactNode[]): string {
	if (! Array.isArray(value)) return slug(value?.toString() ?? '');
	return slug(
		value.map((node) => {
			if (typeof node === 'string' ||
				typeof node === 'number' ||
				typeof node === 'boolean'
			) return node;

			if (typeof node === 'object' && node !== null && 'props' in node) {
				return makeSlug(node.props.children)
			}

			return ''
		}).join('')
	)
}