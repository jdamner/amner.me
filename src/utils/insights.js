import { track, parameters, init } from 'insights-js';

const pageView = () => {
    event('page-view');
}

const event = (eventName, eventProperties = {}) => {
    init('HbSIygPWVvMzXFRP');
    track({
        id: eventName,
        parameters: {
            ...parameters,
            ...eventProperties,
        },
    });
}

export { pageView, event };