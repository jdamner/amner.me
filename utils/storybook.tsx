import { faker } from "@faker-js/faker";
import { MdFile, PostLinkType } from "../types";
import matter from "gray-matter";
import React from "react";

type MdAttributes = Record<string, string>;

faker.seed(123);

export const generateMdFile = <T extends MdAttributes>(
  attributes: T,
  content?: string,
): Omit<MdFile<string>, "data"> & { data: T } => {
  return matter(generateMD(attributes, content)) as Omit<
    MdFile<string>,
    "data"
  > & { data: T };
};

export const generateMdContent = (): string => {
  return (
    `# ${faker.lorem.words(3)}\n` +
    Array.from({ length: 3 }, () => {
      return (
        `## ${faker.lorem.words(2)}\n` +
        faker.lorem.paragraph() +
        "\n" +
        Array.from({ length: 3 }, () => {
          return (
            `### ${faker.lorem.words(2)}\n` + `${faker.lorem.paragraph()}\n`
          );
        }).join("")
      );
    }).join("")
  );
};

export const generatePostLink = (): PostLinkType => {
  return {
    title: faker.lorem.words(3),
    thumbnail: faker.image.url(),
    slug: faker.lorem.slug(),
  };
};

export const generatePostLinks = (count: number): PostLinkType[] => {
  return Array.from({ length: count }, () => generatePostLink());
};

export const generateMD = (
  attributes: MdAttributes,
  content?: string,
): string => {
  const frontMatter = generateMDFrontMatter(attributes);
  return `${frontMatter}\n${content ?? generateMdContent()}`;
};

export const generateMDFrontMatter = (attributes: MdAttributes): string => {
  return (
    "---\n" +
    Object.keys(attributes)
      .map((key) => `${key}: ${attributes[key]}`)
      .join("\n") +
    "\n---"
  );
};

export const generatePlaceholderContent = (
  size: number = 5,
): React.ReactNode => {
  return (
    <>
      {Array.from({ length: size }).map((_, i) => (
        <p key={i}>{faker.lorem.paragraph(i + 1)}</p>
      ))}
    </>
  );
};
