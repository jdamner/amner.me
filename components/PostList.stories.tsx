import type { Meta, StoryObj } from "@storybook/react";

import PostList from "./PostList";
import { generatePostLinks } from "../utils/storybook";

type Story = StoryObj<typeof PostList>;

export default {
  title: "Components/PostList",
  component: PostList,
} as Meta;

export const Default: Story = {
  args: {
    posts: generatePostLinks(10),
  },
};
