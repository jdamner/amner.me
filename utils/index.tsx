/* External Deps */
import { readFileSync, readdirSync } from "fs";
import { resolve } from "path";
import matter from "gray-matter";
import { slug } from "github-slugger";

/* Components */
import { Markdown } from "@/components/Markdown";

/* Constants */
const postsDirectory: string = resolve("content/blog/");
const servicesDirectory: string = resolve("content/services");
const cvDirectory: string = resolve("content/cv");
const projectsDirectory: string = resolve(cvDirectory, "projects");
const referencesDirectory: string = resolve(cvDirectory, "references");
const employmentDirectory: string = resolve(cvDirectory, "employment");
const educationDirectory: string = resolve(cvDirectory, "education");

type MatterData = ReturnType<typeof matter>["data"];
type AttributeValidator<T extends MatterData> = (data: MatterData) => data is T;

const defaultAttributeValidator = <T extends MatterData>(
  data: MatterData,
): data is T => {
  return typeof data === "object" && data !== null;
};

/**
 * Gets all files in a directory
 */
const getAllFrom = <T extends MatterData>(
  directory: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  readdirSync(directory).map((filename) =>
    readMdFile<T>(filename, directory, attributeValidator),
  );

/**
 * Reads a markdown file
 */
const readMdFile = <T extends MatterData>(
  filename: string,
  directory: string = "content/",
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  parseMarkdown<T>(
    readFileSync(`${directory}/${filename}`, "utf8"),
    filename,
    attributeValidator,
  );

/**
 * Parses markdown file contents
 */
export const parseMarkdown = <T extends MatterData>(
  fileContents: string,
  filename: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) => {
  const { content, data } = matter(fileContents);

  if (!attributeValidator(data)) {
    console.log(data);
    throw new Error(`Invalid attributes in ${filename}` + JSON.stringify(data));
  }

  return {
    data,
    slug: slug(filename.replace(/\.mdx$/, "")),
    content: <Markdown>{content}</Markdown>,
  };
};

/**
 * Gets all Posts
 */
const getAllPosts = () =>
  getAllFrom<WithTitle & WithThumbnail & WithDate>(
    postsDirectory,
    (data): data is WithTitle & WithThumbnail & WithDate =>
      hasTitle(data) && hasThumbnail(data) && hasDate(data),
  );

export const useAllServices = () =>
  getAllFrom<WithTitle>(servicesDirectory, hasTitle);

export const getAllProjects = () =>
  getAllFrom<WithTitle & WithLink>(
    projectsDirectory,
    (data): data is WithTitle & WithLink => hasTitle(data) && hasLink(data),
  );

export const useAllReferences = () =>
  getAllFrom<WithTitle & WithWebsite & WithPhone & WithEmail>(
    referencesDirectory,
    (data): data is WithTitle & WithWebsite & WithPhone & WithEmail =>
      hasTitle(data) && hasWebsite(data) && hasPhone(data) && hasEmail(data),
  );

export const useAllEmployment = () =>
  getAllFrom<
    WithTitle & WithSubtitle & WithDateString & WithRole & WithWebsite
  >(
    employmentDirectory,
    (
      data,
    ): data is WithTitle &
      WithSubtitle &
      WithDateString &
      WithRole &
      WithWebsite =>
      hasTitle(data) &&
      hasSubtitle(data) &&
      hasDateString(data) &&
      hasRole(data) &&
      hasWebsite(data),
  );

export const useAllEducation = () =>
  getAllFrom<WithTitle & WithDateString>(
    educationDirectory,
    (data): data is WithTitle & WithDateString =>
      hasTitle(data) && hasDateString(data),
  );

export const getPost = (slug: string) =>
  getAllPosts().find((post) => post.slug === slug);

export const getAllPostLinks = () =>
  getAllPosts()
    .map((post) => ({
      title: post.data.title,
      thumbnail: post.data.thumbnail,
      slug: post.slug,
      date: post.data.date,
    }))
    .sort((a, b) => (a.date > b.date ? -1 : 1));

type WithTitle = { title: string };
type WithSubtitle = { subtitle: string };
type WithLink = { link: string };
type WithThumbnail = { thumbnail: string };
type WithDate = { date: Date };
type WithDateString = { date: string };
type WithPhone = { phone: string };
type WithWebsite = { website: string };
type WithEmail = { email: string };
type WithRole = { role: string };

const hasTitle = (data: MatterData): data is WithTitle =>
  defaultAttributeValidator(data) &&
  "title" in data &&
  typeof data.title === "string";

const hasSubtitle = (data: MatterData): data is WithSubtitle =>
  "subtitle" in data && typeof data.subtitle === "string";

const hasLink = (data: MatterData): data is WithLink =>
  "link" in data && typeof data.link === "string";

const hasThumbnail = (data: MatterData): data is WithThumbnail =>
  "thumbnail" in data && typeof data.thumbnail === "string";

const hasDate = (data: MatterData): data is WithDate =>
  "date" in data && data.date instanceof Date;

const hasDateString = (data: MatterData): data is WithDateString =>
  "date" in data && typeof data.date === "string";

const hasPhone = (data: MatterData): data is WithPhone =>
  "phone" in data && typeof data.phone === "string";

const hasWebsite = (data: MatterData): data is WithWebsite =>
  "website" in data && typeof data.website === "string";

const hasEmail = (data: MatterData): data is WithEmail =>
  "email" in data && typeof data.email === "string";

export const hasRole = (data: MatterData): data is WithRole =>
  "role" in data && typeof data.role === "string";
