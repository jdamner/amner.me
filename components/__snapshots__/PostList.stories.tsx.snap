// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/PostList Default smoke-test 1`] = `
<div class="bg-black text-orange-100  font-sans">
  <div>
    <h2 class="tracking-tight uppercase text-6xl md:text-7xl lg:text-8xl text-orange-100 mb-5 bebas-neue-normal-400">
      Blog Posts
    </h2>
    <ul>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/ipsa-super-curo"
        >
          subito circumvenio calamitas
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/cruciamentum-conicio-suspendo"
        >
          vomer stillicidium denique
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/taceo-auctor-atque"
        >
          decens adulatio cubicularis
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/ulciscor-surculus-perferendis"
        >
          exercitationem exercitationem reiciendis
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/calco-claudeo-quos"
        >
          suppono commodi conturbo
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/depulso-cursim-color"
        >
          aliquam damno curvo
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/desipio-quia-ancilla"
        >
          cursim varietas vinco
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/casso-denuncio-voro"
        >
          comes cuppedia usitas
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/trepide-pax-impedit"
        >
          earum placeat animi
        </a>
      </li>
      <li class="mb-1">
        <a class="inline-block border-2 border-orange-600 bg-orange-600 text-black uppercase text-2xl font-black px-4 py-2 hover:bg-orange-100 hover:text-orange-600 hover:border-orange-600 bebas-neue-normal-400"
           href="/stabilis-utique-dolore"
        >
          conicio cognomen cur
        </a>
      </li>
    </ul>
  </div>
</div>
`;
